docs/middleware.rst

```diff
@@ -495,3 +499,101 @@ A legacy transaction still works in the same way as it did before EIP-1559 was i
    ... }
    >>> w3.eth.send_transaction(legacy_transaction)
+
+
+Create Middleware
+-----------------
+
+To write your own middleware, create a class and extend from the base ``Web3Middleware``
+class, then override only the parts of the middleware that make sense for your use case.
+
+.. note:: The Middleware API borrows from the Django middleware API introduced
+          in version 1.10.0.
+
+If all you need is to modify the params before a request is made, you can override
+the ``request_processor`` method, make the necessary tweaks to the params, and pass the
+arguments to the next element in the middleware stack. Need to do some processing on the
+response? Override the ``response_processor`` method and return the modified response.
+
+The pattern:
+
+.. code-block:: python
+
+    from web3.middleware import Web3Middleware
+
+    class FancyMiddleware(Web3Middleware):
+
+        def request_processor(self, method, params):
+            # Pre-request processing goes here before passing to the next middleware.
+            return (method, params)
+
+        def response_processor(self, method, response):
+            # Response processing goes here before passing to the next middleware.
+            return response
+
+        # If your provider is asynchronous, override the async methods instead:
+
+        async def async_request_processor(self, method, params):
+            # Pre-request processing goes here before passing to the next middleware.
+            return (method, params)
+
+        async def async_response_processor(self, method, response):
+            # Response processing goes here before passing to the next middleware.
+            return response
+
+
+If you wish to prevent making a call under certain conditions, you can override the
+``wrap_make_request`` method. This allows for defining pre-request processing,
+skipping or making the request under certain conditions, as well as response
+processing before passing it to the next middleware.
+
+
+.. code-block:: python
+
+    from web3.middleware import Web3Middleware
+
+    class FancyMiddleware(Web3Middleware):
+
+        def wrap_make_request(self, make_request):
+            def middleware(method, params):
+                # pre-request processing goes here
+                response = make_request(method, params)  # make the request
+                # response processing goes here
+                return response
+
+            return middleware
+
+        # If your provider is asynchronous, override the async method instead:
+
+        async def async_wrap_make_request(self, make_request):
+            async def middleware(method, params):
+                # pre-request processing goes here
+                response = await make_request(method, params)
+                # response processing goes here
+                return response
+
+            return middleware
+
+
+Adding Created Middleware
```

Comment:
This seems like it can be condensed with the API docs above, especially the description of how the API is exposed. Feels like that should be in the API section instead. There's no difference between adding custom or default middleware. Maybe a smaller blip rather than a whole section here or linking to the API section above to add / remove middleware?

If small blip, something more succinct will do here rather than repeating the API steps for `add()`

Reply:
simplified 🤝 i think having a reference example that includes the custom module import is valuable.